### PR TITLE
Fix compiler flags not used for some tests

### PR DESCRIPTION
**Pull request by ePirat**

### CODE REVIEW

**Comment by ePirat** [comment]
I can't understand the test failures.

File: mesonbuild/compilers/clike.py [hunk 1]
```diff
@@ -384,7 +384,7 @@ def _get_basic_compiler_args(self, env, mode):
         # Select a CRT if needed since we're linking
         if mode == 'link':
             args += self.get_linker_debug_crt_args()
-        if mode in {'compile', 'preprocess'}:
+        if mode in {'compile', 'preprocess', 'link'}:
```
Comment (Ericson2314) [inline mesonbuild/compilers/clike.py]: the `elif mode == 'link':` below is now dead code. Relatedly, is `mode in {'compile', 'preprocess', 'link'}:` every not true?

Comment (dcbaker) [reply mesonbuild/compilers/clike.py]: I think this is the problem, drop the `if mode in ...` check and make that code unconditional, and then replace the `elif` with `if`.

File: mesonbuild/compilers/clike.py [hunk 2]
```diff
@@ -384,16 +384,17 @@ def _get_basic_compiler_args(self, env, mode):
...
+
+        # Add CFLAGS/CXXFLAGS/OBJCFLAGS/OBJCXXFLAGS and CPPFLAGS from the env
+        sys_args = env.coredata.get_external_args(self.for_machine, self.language)
+        # Apparently it is a thing to inject linker flags both
+        # via CFLAGS _and_ LDFLAGS, even though the former are
+        # also used during linking. These flags can break
+        # argument checks. Thanks, Autotools.
+        cleaned_sys_args = self.remove_linkerlike_args(sys_args)
+        args += cleaned_sys_args
+
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/clike.py]: __[Flake8]__ [W293] blank line contains whitespace ([link](https://sider.review/gh/repos/19784232/pulls/5483#issue-4249954))

<sub>posted by Sider</sub>

**Comment by Ericson2314** [comment]
CC @jmesmon, who has poured over this code now, on the latest errors.

**Comment by jpakkane** [comment]
Ci is fixed apart from VS2019, please rebase to rerun tests.

**Comment by ePirat** [comment]
Ok… So looking at the failed test it seems I somehow end up with some arguments twice… Whoops…

**Comment by jmesmon** [comment]
The addition of the `env` to all modes looks fine. Switching to using `_build_wrapper()` though does currently mean that the `get_compile_only_args()` (and their friends `get_always_args()` and `get_preprocess_only_args()`) won't get added to the flags anymore (normally, `compile()` adds them).

**Comment by ePirat** [comment]
Ok, do you have an idea how to fix this? The whole point of changing it was because flags did not get added when using `compile()` for some reason. That said this all is super confusing and I am clueless what the purpose of the `get_compile_only_args()`, `get_always_args()` and `get_preprocess_only_args()` even is as this is not really easy to understand :/

**Comment by Ericson2314** [comment]
https://github.com/mesonbuild/meson/blob/ccccb874cd48c6442146e400f5cb7fc6d288a53c/mesonbuild/compilers/clike.py#L436 I think `_build_wrapper` calls `compile`, so all should be good? This + #5510 looks simpler and more correct to me. :tada: Thanks, you both!

**Comment by jmesmon** [comment]
Yes, I was mistaken about `_build_wrapper()` and `compile`'s relationship. This looks good.

**Comment by jpakkane** [comment]
There has been related fixes and the CI run is old with the failing messages. Please rebase to fix it.

**Comment by jpakkane** [comment]
```
Sanity check compiler command line: cl /MDd foo bar one two /nologo /showIncludes D:\a\1\s\tmpenwrqyw_\meson-private\sanitycheckc.c /FeD:\a\1\s\tmpenwrqyw_\meson-private\sanitycheckc.exe
Sanity check compile stdout:
sanitycheckc.c
LINK : fatal error LNK1181: cannot open input file 'foo.obj'
```

and also on Travis:

```
Sanity check compiler command line: clang foo bar one two -pipe -D_FILE_OFFSET_BITS=64 /root/tmpaa3fmcef/meson-private/sanitycheckc.c -o /root/tmpaa3fmcef/meson-private/sanitycheckc.exe
Sanity check compile stdout:
-----
Sanity check compile stderr:
clang: error: no such file or directory: 'foo'
clang: error: no such file or directory: 'bar'
clang: error: no such file or directory: 'one two'
```

**Comment by ePirat** [comment]
@jpakkane I do not understand what the failing test is or tries to do there, so I am absolutely clueless, unfortunately, what to do about it.

**Comment by jpakkane** [comment]
The relevant test line seems to be this:

```
self.init(testdir, extra_args=['-Dc_args=foo bar "one two"'])
```

This equals the following Meson invocation:

```
meson <other options> '-Dc_args=foo bar "one two"'
```

Since c_args is a string but it needs to be able to handle spaces it is split according to Unix commandline split rules. Thus it becomes `['foo', 'bar', 'one two']`. Now it seems that these flags were not used in the sanity check compilation test earlier but now they are used leading to the failure.

So the question is whether `c_args` should be used in sanity tests when using the build compiler. The old assumption was that the system compiler should be invokable and runnable without any extra args so the flags were not used. But on the other hand a case can be made that those args should be used in the sanity check so erroneous flags like these are detected earlier.

I'm leaning slightly towards the first choice to maintain backwards compatibility.

**Comment by ePirat** [comment]
Autotools uses them in the sanity check, it has the useful effect that wrong flags fail early instead of causing harder to debug failures later… But I can see why it might be desired to have the other behavior…

**Comment by jpakkane** [comment]
Well you can do that, you just need to change the arguments to `-Done -Dtwo "-Done two"`.

**Comment by ePirat** [comment]
@jpakkane How did this work before? I mean, how did it only apply the args when not doing the compiler sanity check? There seems no way to tell apart the invocations inside the relevant method I changed?

**Comment by jpakkane** [comment]
IIRC the old code did not take _any_ outside arguments. It built the command line manually from scratch. Dunno how it worked between ye olden times and the latest release.

**Comment by Ericson2314** [comment]
I would just change the test. I definitely like the "fail fast" benefit and build vs host, native vs cross consistency.

**Comment by ePirat** [comment]
I am a bit confused reading https://mesonbuild.com/Release-notes-for-0-51-0.html#sanity-checking-compilers-with-user-flags which says that the behavior introduced by my patch should have already been the case in 0.51.0?

**Comment by Ericson2314** [comment]
@ePriat yes it was my intention that this be the case, but the very complicated indirection made led me to screw up. Perhaps this and #5510 should be backported to the release branch.

**Comment by jpakkane** [comment]
They are both already marked for the point release. In fact this is the last one and once it is merged, we can do the point release.

**Comment by jpakkane** [comment]
I fixed the test in #5584.

**Comment by jpakkane** [comment]
There is a second failing thing here, too:

```
======================================================================
FAIL: test_compiler_check_flags_order (__main__.LinuxlikeTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "run_unittests.py", line 4478, in test_compiler_check_flags_order
    self.assertEqual(Oargs, [Oflag, '-O0'])
AssertionError: Lists differ: ['-O3', '-O3', '-O0'] != ['-O3', '-O0']
First differing element 1:
'-O3'
'-O0'
First list contains 1 additional elements.
First extra element 2:
'-O0'
- ['-O3', '-O3', '-O0']
?         -------
+ ['-O3', '-O0']
```

**Comment by ePirat** [comment]
@jpakkane Thanks. Yes that second test was actually the one that I saw the first time failing and was clueless about it. I suspect somehow we end up with some flags duplicated? But not sure how…

**Comment by ePirat** [comment]
(I invite anyone with more knowledge of the code to just take my PR and change it, I don't care about authorship either, after all the change is trivial)

**Comment by ePirat** [comment]
@jpakkane Is this fixed by a different PR? If not, I can have another look. Is there an easy way to just run the specific failing test locally without all the other tests?

**Comment by jpakkane** [comment]
#5584.

> Is there an easy way to just run the specific failing test locally without all the other tests?

The standard Python unittest way:

    ./run_unittests.py LinuxlikeTests.test_compiler_check_flags_order

**Comment by Ericson2314** [comment]
@ePirat I made https://docs.pytest.org/en/latest/ work on the unit tests, which allows handy things like running all previously failing tests.

**Pull request closed without merging**